File: backend/app/services/user_service.py
from app.models.user import User
from app.models.role import Role
from app import db


class UserService:
    @staticmethod
    def get_user_by_email(email):
        return User.query.filter_by(email=email).first()

    @staticmethod
    def create_user(first_name, last_name, email, phone_number, password, role_id):
        new_user = User(
            first_name=first_name,
            last_name=last_name,
            email=email,
            phone_number=phone_number,
            role_id=role_id
        )
        new_user.set_password(password)
        db.session.add(new_user)
        db.session.commit()
        return new_user

    @staticmethod
    def get_all_users():
        return User.query.all()

    @staticmethod
    def get_user_by_id(user_id):
        return User.query.get(user_id)

    @staticmethod
    def delete_user(user_id):
        user = User.query.get(user_id)
        if user:
            db.session.delete(user)
            db.session.commit()
            return True
        return False

    @staticmethod
    def update_user(user_id, data):
        user = User.query.get(user_id)
        if user:
            user.first_name = data.get('first_name', user.first_name)
            user.last_name = data.get('last_name', user.last_name)
            user.email = data.get('email', user.email)
            user.phone_number = data.get('phone_number', user.phone_number)
            user.role_id = data.get('role_id', user.role_id)
            db.session.commit()
            return user
        return None

    @staticmethod
    def get_user_roles():
        return Role.query.all()

File: frontend/lib/core/services/user_service.dart
import 'dart:convert';
import 'package:frontend/data/repositories/secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:frontend/core/constants/constants.dart';
import 'package:frontend/data/models/user_info_model.dart';

class UserService {
  final String baseUrl = AppConstants.baseUrl;

  Future<List<UserDetail>> getAllUsers() async {
    final url = Uri.parse('$baseUrl/users');
    final response = await http.get(url);

    if (response.statusCode == 200) {
      final List<dynamic> usersJson = jsonDecode(response.body);
      return usersJson.map((json) => UserDetail.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load users');
    }
  }

  Future<UserDetail> getUserbyId(int user_id) async {
    final url = Uri.parse('$baseUrl/users/$user_id');
    final token = await SecureStorage().read("auth_token");
    final response = await http.get(
      url,
      headers: {'Authorization': 'Bearer $token'}, 
    );

    if (response.statusCode == 200) {
      final userJson = jsonDecode(response.body);
      return UserDetail.fromJson(userJson);
    } else {
      throw Exception('Failed to get user');
    }
  }

  Future<UserDetail> updateUser(UserDetail userDetail) async {
    final response = await http.put(
      Uri.parse('$baseUrl/users/${userDetail.userId}'),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'firstName': userDetail.firstName,
        'lastName': userDetail.lastName,
        'email': userDetail.email,
        'phoneNumber': userDetail.phoneNumber,
        'roleId': userDetail.roleId.toString(),
      }),
    );

    if (response.statusCode == 200) {
      return UserDetail.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to update user');
    }
  }
}
